File: matlab/task2_0.m
% 2. Reproduza, utilizando o MATLAB e outra linguagem de programação de sua
% escolha, o exemplo apresentado na página 3 deste documento
% (ImageProcessing Toolbox User'sGuide - UsingMedianFiltering). Repita, no
% MATLAB, para ruído 'gaussian' e 'speckle', explicando o que são estes
% tipos de ruído.Discuta o experimento no relatório
function task2_0()
    assetsDir = ['..' filesep 'assets' filesep];
    filePath = strcat(assetsDir, 'lenna.png');
    
    I = imread(filePath);
    I = rgb2gray(I);
    
    % Apply Noise
    SP = imnoise(I,'salt & pepper',0.02);
    
    % Filter the noisy image with an averaging filter and display the results.
    SPAvrg = filter2(fspecial('average',3),SP)/255;
    
    % Now use a median filter to filter the noisy image and display the results. Notice
    % that medfilt2 does a better job of removing noise, with less blurring of edges.
    SPMed = medfilt2(SP,[3 3]);    
    
    % Gaussian Noise
    G = imnoise(I,'gaussian',0.02);
    GAvrg = filter2(fspecial('average',3),G)/255;
    GMed = medfilt2(G,[3 3]);   
    
    % Speckle Noise
    S = imnoise(I,'speckle',0.02);
    SAvrg = filter2(fspecial('average',3),S)/255;
    SMed = medfilt2(S,[3 3]);   
    
    % Ploting
    fig = figure(1); set (fig, 'Units', 'normalized', 'Position', [0,0,1,1]);    
    %subplot(4,3,1), imshow(I,[]), title('Original image')
    subplot(1,3,1), imshow(SP,[]), title('Salt&Pepper Noise')
    subplot(1,3,2), imshow(SPAvrg,[]), title('Average Filter')
    subplot(1,3,3), imshow(SPMed,[]), title('Median Filter')
    
    fig = figure(2); set (fig, 'Units', 'normalized', 'Position', [0,0,1,1]);   
    subplot(1,3,1), imshow(G,[]), title('Gaussian Noise')
    subplot(1,3,2), imshow(GAvrg,[]), title('Average Filter')
    subplot(1,3,3), imshow(GMed,[]), title('Median Filter')
    
    fig = figure(3); set (fig, 'Units', 'normalized', 'Position', [0,0,1,1]);   
    subplot(1,3,1), imshow(S,[]), title('Speckle Noise')
    subplot(1,3,2), imshow(SAvrg,[]), title('Average Filter')
    subplot(1,3,3), imshow(SMed,[]), title('Median Filter')
    
end

File: python/task2_0.py
# coding=UTF-8
# 2.0

import cv2
import utils
import random
import numpy as np


def main():
    image = utils.load_image('lenna.png')

    sp_image = _salt_pepper(image, 0.05)
    utils.display_single_image('Salt&Pepper Noise', sp_image)    

    print _mean_filter(3)

    # filter1 = _filter(sp_image, [[1/9, 1/9, 1/9], [1/9, 1/9, 1/9], [1/9, 1/9, 1/9]])
    # utils.display_single_image('Mean Filter', filter1)    

    # filter2 = _filter(sp_image, [[0.0, 0.125, 0.0], [0.125, 0.25, 0.125], [0.0, 0.125, 0.0]])
    # utils.display_single_image('Median Filter', filter2)    

    utils.wait_key_and_destroy_windows()


def _salt_pepper(image, p):
    output = np.zeros(image.shape,np.uint8)    
    for i in range(image.shape[0]):
        for j in range(image.shape[1]):
            rdn = random.random()            
            if rdn < p:                
                rdn = random.random()
                output[i][j] = (rdn>0.5)*255                            
            else:
                output[i][j] = image[i][j]
    return output

if __name__ == "__main__":
    main()
